### PR TITLE
annotate: use memchr to speed up line counting

### PR DESCRIPTION
**Pull request by camelid**
'jj file annotate' is often much slower than 'git blame', which can make it frustrating to use (cc #4878). This change speeds up 'jj file annotate' significantly by using memchr. When testing on a file in my local rust-lang/rust checkout with a long history, I found a speedup of about 35%.

Before:

    ❯ hyperfine "~/code/jj/target/release/jj --no-pager file annotate src/librustdoc/clean/mod.rs"
    Benchmark 1: ~/code/jj/target/release/jj --no-pager file annotate src/librustdoc/clean/mod.rs
      Time (mean ± σ):     18.174 s ±  0.348 s    [User: 18.171 s, System: 0.207 s]
      Range (min … max):   17.545 s … 18.836 s    10 runs

After:

    ❯ hyperfine "~/code/jj/target/release/jj --no-pager file annotate src/librustdoc/clean/mod.rs"
    Benchmark 1: ~/code/jj/target/release/jj --no-pager file annotate src/librustdoc/clean/mod.rs
      Time (mean ± σ):     13.472 s ±  0.141 s    [User: 13.471 s, System: 0.206 s]
      Range (min … max):   13.291 s … 13.714 s    10 runs

LLM disclaimer: I used an LLM to investigate this performance problem and find a fix, but I manually checked and tweaked the code and ran the above performance evaluation.

# Checklist

If applicable:

- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (`README.md`, `docs/`, `demos/`)
- [ ] I have updated the config schema (`cli/src/config-schema.json`)
- [x] I have added/updated tests to cover my changes
- [x] I fully understand the code that I am submitting (what it does,
      how it works, how it's organized), including any code drafted by an LLM.
- [x] For any prose generated by an LLM, I have proof-read and copy-edited with
      an eye towards deleting anything that is irrelevant, clarifying anything
      that is confusing, and adding details that are relevant. This includes,
      for example, commit descriptions, PR descriptions, and code comments.

### CODE REVIEW

**Comment by josephlou5** [comment]
Small nit: I think it could be worth mentioning the issue in the commit description as well, so that there's a backlink in the repo history itself rather than only on the GitHub PR.

File: lib/src/annotate.rs [hunk 1]
```diff
@@ -423,21 +423,30 @@ fn copy_same_lines_with(
...
         }
     }
 }
 
+/// Counts the number of lines in `text`, where the last line may not end with
+/// a newline character. Equivalent to `text.split_inclusive(|b| *b ==
+/// b'\n').count()`, but faster.
+fn count_lines(text: &[u8]) -> usize {
+    let newlines = memchr::memchr_iter(b'\n', text).count();
+    let end_is_not_newline = text.last().is_some_and(|b| *b != b'\n');
```
Comment (yuja) [inline lib/src/annotate.rs:446]: Can you add a few edge-case tests for `""`, `"x"`, `"x\n"`, ...?

**Comment by camelid** [comment]
Done and done.

**Review by yuja — APPROVED** [review]

**Review by thoughtpolice — APPROVED** [review]

**Comment by PhilipMetzger** [comment]
Austin you need to land @senekor's clippy fix first before landing this. The updated clippy reports a new error.
